tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/PublicAssetRepository.scala

```diff
@@ -0,0 +1,117 @@
+package com.linagora.tmail.james.jmap.publicAsset
+
+import java.io.{ByteArrayInputStream, InputStream}
+
+import com.google.common.collect.{HashBasedTable, Table, Tables}
+import com.google.common.io.CountingInputStream
+import com.linagora.tmail.james.jmap.publicAsset.ImageContentType.ImageContentType
+import com.linagora.tmail.james.jmap.publicAsset.MemoryPublicAssetRepository.PublicAssetMetadata
+import org.apache.james.blob.api.{BlobId, BlobStore, BucketName}
+import org.apache.james.core.Username
+import org.apache.james.jmap.api.model.IdentityId
+import org.apache.james.jmap.api.model.Size.Size
+import org.apache.james.util.ReactorUtils
+import org.reactivestreams.Publisher
+import reactor.core.scala.publisher.{SFlux, SMono}
+
+import scala.jdk.CollectionConverters._
+
+trait PublicAssetRepository {
+  def create(username: Username, creationRequest: PublicAssetCreationRequest): Publisher[PublicAssetStorage]
+
+  def update(username: Username, id: PublicAssetId, identityIds: Set[IdentityId]): Publisher[Void]
+
+  def remove(username: Username, id: PublicAssetId): Publisher[Void]
+
+  def revoke(username: Username): Publisher[Void]
+
+  def get(username: Username, ids: Set[PublicAssetId]): Publisher[PublicAssetStorage]
+
+  def get(username: Username, id: PublicAssetId): Publisher[PublicAssetStorage] = get(username, Set(id))
+
+  def list(username: Username): Publisher[PublicAssetStorage]
+
+}
+
+object MemoryPublicAssetRepository {
+
+  object PublicAssetMetadata {
+    def from(publicAsset: PublicAssetStorage): PublicAssetMetadata =
+      PublicAssetMetadata(
+        publicAsset.id,
+        publicAsset.publicURI,
+        publicAsset.size,
+        publicAsset.contentType,
+        publicAsset.blobId,
+        publicAsset.identityIds)
+  }
+
+  case class PublicAssetMetadata(id: PublicAssetId,
+                                 publicURI: PublicURI,
+                                 size: Size,
+                                 contentType: ImageContentType,
+                                 blobId: BlobId,
+                                 identityIds: Seq[IdentityId]) {
+
+    def asPublicAssetStorage(bytes: Array[Byte]): PublicAssetStorage =
+      PublicAssetStorage(id = id,
+        publicURI = publicURI,
+        size = size,
+        contentType = contentType,
+        blobId = blobId,
+        identityIds = identityIds,
+        content = () => new ByteArrayInputStream(bytes))
+  }
+}
+
+class MemoryPublicAssetRepository(val blobStore: BlobStore) extends PublicAssetRepository {
```

Comment:
Now that we store things in the blob store we also need to list references for the GC otherwize blobs gonna be deleted in the background!